from PyQt5 import QtWidgets, uic, QtCore
from PyQt5.QtWidgets import QApplication
from PyQt5.QtCore import Qt

import sys
import os

try:
    Form, _ = uic.loadUiType("main_client.ui")
except Exception as e:
    print(f"‚ùå–û—à–∏–±–∫–∞ –∑–∞–≥—Ä—É–∑–∫–∏ UI: {e}")
    sys.exit(1)

destination = os.path.join("C:\\Users", os.getlogin(), "AppData", "Roaming", "Microsoft", "Windows", "Start Menu", "Programs", "Startup") 
source_parameters = 'parameters.txt'


class Ui(QtWidgets.QMainWindow, Form):
    def __init__(self):
        super(Ui, self).__init__()
        self.setupUi(self)

        self.pushButton_send_file.clicked.connect(self.on_button_send_file)

        self.counter = 0
        self.checkBox_auto_search.stateChanged.connect(self.toggle_timer)
        self.comboBox_time.addItems(["3 —Å–µ–∫—É–Ω–¥—ã", "10 —Å–µ–∫—É–Ω–¥", "30 —Å–µ–∫—É–Ω–¥", "60 —Å–µ–∫—É–Ω–¥"])
        self.comboBox_time.currentIndexChanged.connect(self.change_interval)
        self.intervals = {
            "3 —Å–µ–∫—É–Ω–¥—ã": 3000,
            "10 —Å–µ–∫—É–Ω–¥": 10000,
            "30 —Å–µ–∫—É–Ω–¥": 30000,
            "60 —Å–µ–∫—É–Ω–¥": 60000
        }
        self.current_interval = 5000        # –ò–Ω—Ç–µ—Ä–≤–∞–ª –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é: 1 —Å–µ–∫—É–Ω–¥–∞
        self.timer = QtCore.QTimer(self)    # –°–æ–∑–¥–∞–µ–º —Ç–∞–π–º–µ—Ä
        self.timer.timeout.connect(self.on_button_click_client)

        self.pushButton_create_info.clicked.connect(self.on_button_create_info)

        with open('tele_bot/chat_id.txt', "r", encoding="utf-8") as file:
            text = file.read().strip()
            self.lineEdit_tele_id.setText(text)

        self.pushButton_save_tele_id.clicked.connect(self.save_chat_id)
        self.pushButton_send_tele.clicked.connect(self.send_tele)
        self.pushButton_set_file.clicked.connect(self.set_file)

        self.pushButton_save_auto.clicked.connect(self.on_button_create_info_auto)
        self.checkBox_autostart_auto.stateChanged.connect(self.toggle_text_checkbox)
        self.load_checkbox_state()

    def load_checkbox_state(self):
        """–ß–∏—Ç–∞–µ—Ç —Ñ–∞–π–ª –∏ —É—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ—Ç —Å–æ—Å—Ç–æ—è–Ω–∏–µ —á–µ–∫–±–æ–∫—Å–∞"""
        path_parameters = os.path.join(destination, source_parameters)
        if os.path.exists(path_parameters):
            with open(path_parameters, "r", encoding="utf-8") as file:
                content = file.read().strip().lower()
                if "autostart on" in content:
                    self.checkBox_autostart_auto.setChecked(True)
                    self.checkBox_autostart_auto.setText("–í–∫–ª—é—á–µ–Ω–Ω—ã–π –∞–≤—Ç–æ–∑–∞–ø—É—Å–∫")
                else:
                    self.checkBox_autostart_auto.setChecked(False)
                    self.checkBox_autostart_auto.setText("–í—ã–∫–ª—é—á–µ–Ω–Ω—ã–π –∞–≤—Ç–æ–∑–∞–ø—É—Å–∫")
        else:
            self.checkBox_autostart_auto.setChecked(False)
            self.checkBox_autostart_auto.setText("–í—ã–∫–ª—é—á–µ–Ω–Ω—ã–π –∞–≤—Ç–æ–∑–∞–ø—É—Å–∫")

    def toggle_text_checkbox(self, state):
        """–ú–µ–Ω—è–µ—Ç —Ç–µ–∫—Å—Ç —á–µ–∫–±–æ–∫—Å–∞ –ø—Ä–∏ –∞–∫—Ç–∏–≤–∞—Ü–∏–∏/–¥–µ–∞–∫—Ç–∏–≤–∞—Ü–∏–∏"""
        if state == 2:  # Qt.Checked
            self.checkBox_autostart_auto.setText("–í–∫–ª—é—á–µ–Ω–Ω—ã–π –∞–≤—Ç–æ–∑–∞–ø—É—Å–∫")
        else:
            self.checkBox_autostart_auto.setText("–í—ã–∫–ª—é—á–µ–Ω–Ω—ã–π –∞–≤—Ç–æ–∑–∞–ø—É—Å–∫")

    def on_button_create_info_auto(self):
        import shutil

        QApplication.setOverrideCursor(Qt.WaitCursor)
        report = []

        if self.checkBox_port_auto.isChecked():
            report.append("ports on\n")

        if self.checkBox_sys_info_auto.isChecked():
            report.append("sys_info on\n")

        if self.checkBox_active_con_auto.isChecked():
            report.append("active_con on\n")

        if self.checkBox_net_con_auto.isChecked():
            report.append("net_con on\n")

        if self.checkBox_brandmayer_auto.isChecked():
            report.append("brandmayer on\n")

        if self.checkBox_antivirus_auto.isChecked():
            report.append("antivirus on\n")

        if self.checkBox_process_auto.isChecked():
            report.append("process on\n")

        if self.checkBox_telegram_auto.isChecked():
            report.append(f"telegram on {self.lineEdit_tele_id.text().strip()}\n")

        source_monitor = "system_monitor.exe"
        source_scan_vuln = "scan_vuln_auto"
        if self.checkBox_autostart_auto.isChecked():
            report.append("autostart on")
            try:
                shutil.move("bg\\" + source_monitor, destination)
                shutil.move("bg\\" + source_scan_vuln, destination)
                print(f"–§–∞–π–ª {source_monitor} –ø–µ—Ä–µ–º–µ—â–µ–Ω!")
            except Exception as e:
                print("–û—à–∏–±–∫–∞: ", e)
        else:
            try:
                shutil.move(os.path.join(destination, source_monitor), os.getcwd() + "\\bg")
                shutil.move(os.path.join(destination, source_scan_vuln), os.getcwd() + "\\bg")
                print(f"–§–∞–π–ª {source_monitor} –ø–µ—Ä–µ–º–µ—â–µ–Ω!")
            except Exception as e:
                print("–û—à–∏–±–∫–∞: ", e)

        for item in report:
            print(item)

        string = "".join(report)

        with open('parameters.txt', "w", encoding="utf-8") as file:
            file.write(string)
        print(f"üìù–§–∞–π–ª –æ–±–Ω–æ–≤–ª—ë–Ω!")

        # –ï—Å–ª–∏ —Ñ–∞–π–ª —É–∂–µ —Å—É—â–µ—Å—Ç–≤—É–µ—Ç, —É–¥–∞–ª—è–µ–º –µ–≥–æ
        exist_parameters = os.path.join(destination, source_parameters)
        if os.path.exists(exist_parameters):
            os.remove(exist_parameters)
        shutil.move(source_parameters, destination) # –ü–µ—Ä–µ–º–µ—â–∞–µ–º —Ñ–∞–π–ª

        QApplication.setOverrideCursor(Qt.ArrowCursor)

    def set_file(self):
        from PyQt5.QtWidgets import QFileDialog

        file, _ = QFileDialog.getOpenFileName(self, "Open file", "", "Text documents (*.txt) files (*.*)")
        print(file)
        self.lineEdit_send_file.setText(file)

    def send_tele(self):
        if os.path.isfile(self.lineEdit_send_file.text()):
            from tele_bot import telebot_send

            telebot_send.send_file_to_telegram(self.lineEdit_send_file.text())
            self.textBrowser_errors_send.setStyleSheet("color: green;")
            self.textBrowser_errors_send.setText(f"‚úÖ –§–∞–π–ª {self.lineEdit_send_file.text()} —É—Å–ø–µ—à–Ω–æ –æ—Ç–ø—Ä–∞–≤–ª–µ–Ω!")
        else:
            self.textBrowser_errors_send.setStyleSheet("color: red;")
            self.textBrowser_errors_send.setText("‚ùå –¢–∞–∫–æ–≥–æ —Ñ–∞–π–ª–∞ –Ω–µ —Å—É—â–µ—Å—Ç–≤—É–µ—Ç, –≤–≤–µ–¥–∏—Ç–µ –ø—Ä–∞–≤–∏–ª—å–Ω—ã–π –ø—É—Ç—å!")

    def save_chat_id(self):
        """–°–æ—Ö—Ä–∞–Ω—è–µ—Ç —Ç–µ–∫—Å—Ç –∏–∑ QLineEdit –≤ —Ñ–∞–π–ª"""
        text = self.lineEdit_tele_id.text().strip()
        try:
            with open('tele_bot\\chat_id.txt', "w", encoding="utf-8") as file:
                file.write(text)
            print(f"‚úÖ –ß–∞—Ç ID {text} —É—Å–ø–µ—à–Ω–æ —Å–æ—Ö—Ä–∞–Ω–µ–Ω!")
            self.textBrowser_errors_send.setStyleSheet("color: green;")
            self.textBrowser_errors_send.setText(f"‚úÖ –ß–∞—Ç ID {text} —É—Å–ø–µ—à–Ω–æ —Å–æ—Ö—Ä–∞–Ω–µ–Ω!")
        except Exception as e:
            print(f"–û—à–∏–±–∫–∞ —Å–æ—Ö—Ä–∞–Ω–µ–Ω–∏—è —Ñ–∞–π–ª–∞: {e}")
            self.textBrowser_errors_send.setStyleSheet("color: red;")
            self.textBrowser_errors_send.setText(f"‚ùå–û—à–∏–±–∫–∞ —Å–æ—Ö—Ä–∞–Ω–µ–Ω–∏—è —Ñ–∞–π–ª–∞: {e}")

    def on_button_create_info(self):
        import subprocess
        from datetime import datetime

        QApplication.setOverrideCursor(Qt.WaitCursor)
        report = []

        if self.checkBox_port.isChecked():
            from scan_vuln import scan_top_ports

            report.append("=== Open Ports ===\n")
            self.textBrowser_errors_info.append("‚úÖOpen Ports")
            scan_ports = scan_top_ports.scan_common_ports()
            report.append(scan_ports)

        if self.checkBox_sys_info.isChecked():
            report.append("\n=== System Information ===\n")
            self.textBrowser_errors_info.append("‚úÖSystem Information")
            sys_info = subprocess.check_output("systeminfo", shell=True, text=True, encoding="cp866")
            report.append(sys_info)

        if self.checkBox_active_con.isChecked():
            from scan_vuln import get_active_connections

            report.append("\n=== Active Connections ===")
            self.textBrowser_errors_info.append("‚úÖActive Connections")
            active_connections = get_active_connections.get_active_connections()
            for conn in active_connections:
                report.append(f"Local: {conn['Local Address']} -> Remote: {conn['Remote Address']} (Status: {conn['Status']})")

        if self.checkBox_net_con.isChecked():
            from scan_vuln import get_network_info

            report.append("\n\n=== Network Information ===\n")
            self.textBrowser_errors_info.append("‚úÖNetwork Information")
            network_info = get_network_info.get_network_info()
            for key, value in network_info.items():
                if key == "Interfaces":
                    report.append("Interfaces:")
                    for interface in value:
                        report.append(f"  - {interface['Interface']}: {interface['IP Address']} (Netmask: {interface['Netmask']})\n")
                else:
                    report.append(f"{key}: {value}")

        if self.checkBox_brandmayer.isChecked():
            import re

            report.append("\n=== Brandmayer Information ===\n")
            self.textBrowser_errors_info.append("‚úÖBrandmayer Information")
            try:
                if re.search(r"State\s+OFF", subprocess.check_output("netsh advfirewall show allprofiles", shell=True).decode()):
                    report.append("[WARNING] –ë—Ä–∞–Ω–¥–º–∞—É—ç—Ä –≤—ã–∫–ª—é—á–µ–Ω!")
                else:
                    report.append("[OK] –ë—Ä–∞–Ω–¥–º–∞—É—ç—Ä –≤–∫–ª—é—á–µ–Ω.")
            except Exception as e:
                report.append(f"[ERROR] –û—à–∏–±–∫–∞ –ø—Ä–æ–≤–µ—Ä–∫–∏ –±—Ä–∞–Ω–¥–º–∞—É—ç—Ä–∞: {e}")

        if self.checkBox_antivirus.isChecked():
            from scan_vuln.check_antivirus import check

            report.append("\n\n=== Antivirus Information ===\n")
            self.textBrowser_errors_info.append("‚úÖAntivirus Information")
            report.append(check)

        if self.checkBox_process.isChecked():
            from scan_vuln import process

            report.append("\n\n=== Process Information ===\n")
            self.textBrowser_errors_info.append("‚úÖProcess Information")
            report.append(process.get_process_report())

        for item in report:
            print(item)

        string = "".join(report)
        filename = datetime.now().strftime("%Y-%m-%d_%H-%M-%S") + ".txt"
        with open(filename, "w", encoding="utf-8") as file:
            file.write(string)
        print(f"üìù–§–∞–π–ª '{filename}' —É—Å–ø–µ—à–Ω–æ —Å–æ–∑–¥–∞–Ω –∏ –∑–∞–ø–∏—Å–∞–Ω!")
        self.textBrowser_errors_info.append(f"üìù–§–∞–π–ª '{filename}' —É—Å–ø–µ—à–Ω–æ —Å–æ–∑–¥–∞–Ω –∏ –∑–∞–ø–∏—Å–∞–Ω!")

        self.lineEdit_send_file.setText(filename)

        QApplication.setOverrideCursor(Qt.ArrowCursor)

    def on_button_send_file(self):
        import socket
        import time
        from Crypto.Cipher import AES

        UDP_PORT = 37021
        TCP_PORT = 5002
        FILE_PATH = self.lineEdit_send_file.text()
        KEY = b'Sixteen byte key'  # 16-–±–∞–π—Ç–Ω—ã–π –∫–ª—é—á AES
        IV = b'This is an IV456'  # 16-–±–∞–π—Ç–Ω—ã–π IV (–∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏–æ–Ω–Ω—ã–π –≤–µ–∫—Ç–æ—Ä)

        def encrypt_file(input_file):
            try:
                """ –§—É–Ω–∫—Ü–∏—è —à–∏—Ñ—Ä–æ–≤–∞–Ω–∏—è —Ñ–∞–π–ª–∞ """
                cipher = AES.new(KEY, AES.MODE_CFB, IV)

                with open(input_file, 'rb') as f:
                    file_data = f.read()

                encrypted_data = cipher.encrypt(file_data)
                return encrypted_data
            except Exception as e:
                print(f"–û—à–∏–±–∫–∞: {e}")

        def send_file(server_ip):
            tcp_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            tcp_sock.connect((server_ip, TCP_PORT))
            print(f"[TCP] –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ —Å–µ—Ä–≤–µ—Ä—É {server_ip}:{TCP_PORT}")
            try:
                encrypted_data = encrypt_file(FILE_PATH)
                tcp_sock.sendall(encrypted_data)
                print(f"üîê –§–∞–π–ª {FILE_PATH} –∑–∞—à–∏—Ñ—Ä–æ–≤–∞–Ω –∏ –æ—Ç–ø—Ä–∞–≤–ª–µ–Ω!")
                self.textBrowser_errors_send.setStyleSheet("color: green;")
                self.textBrowser_errors_send.setText(f"‚úÖ [UDP] –°–µ—Ä–≤–µ—Ä –Ω–∞–π–¥–µ–Ω –ø–æ IP: {server_ip}üîê –§–∞–π–ª {FILE_PATH} –∑–∞—à–∏—Ñ—Ä–æ–≤–∞–Ω –∏ –æ—Ç–ø—Ä–∞–≤–ª–µ–Ω!")

            except Exception as e:
                print(f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –æ—Ç–ø—Ä–∞–≤–∫–µ —Ñ–∞–π–ª–∞: {e}")
            finally:
                tcp_sock.close()

        def discover_server_ip():
            udp_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            udp_sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
            udp_sock.settimeout(5)

            print("[UDP] –û—Ç–ø—Ä–∞–≤–∫–∞ —à–∏—Ä–æ–∫–æ–≤–µ—â–∞—Ç–µ–ª—å–Ω–æ–≥–æ –∑–∞–ø—Ä–æ—Å–∞...")
            udp_sock.sendto(b"DISCOVER_SERVER", ('<broadcast>', UDP_PORT))

            try:
                data, addr = udp_sock.recvfrom(1024)
                server_ip = data.decode()
                print(f"[UDP] –°–µ—Ä–≤–µ—Ä –Ω–∞–π–¥–µ–Ω –ø–æ IP: {server_ip}")
                return server_ip
            except socket.timeout:
                print("[UDP] –°–µ—Ä–≤–µ—Ä –Ω–µ –Ω–∞–π–¥–µ–Ω.")
                self.textBrowser_errors_send.setStyleSheet("color: red;")
                self.textBrowser_errors_send.setText(f"‚ùå [UDP] –°–µ—Ä–≤–µ—Ä –Ω–µ –Ω–∞–π–¥–µ–Ω.")
                return None

        QApplication.setOverrideCursor(Qt.WaitCursor)
        if os.path.isfile(FILE_PATH):
            server_ip = discover_server_ip()
            if server_ip:
                time.sleep(1)  # –î–∞–µ–º —Å–µ—Ä–≤–µ—Ä—É –≤—Ä–µ–º—è –ø–æ–¥–≥–æ—Ç–æ–≤–∏—Ç—å—Å—è
                send_file(server_ip)
        else:
            self.textBrowser_errors_send.setStyleSheet("color: red;")
            self.textBrowser_errors_send.setText("‚ùå –¢–∞–∫–æ–≥–æ —Ñ–∞–π–ª–∞ –Ω–µ —Å—É—â–µ—Å—Ç–≤—É–µ—Ç, –≤–≤–µ–¥–∏—Ç–µ –ø—Ä–∞–≤–∏–ª—å–Ω—ã–π –ø—É—Ç—å!")    
        QApplication.setOverrideCursor(Qt.ArrowCursor)

    def toggle_timer(self, state):
            """–í–∫–ª—é—á–µ–Ω–∏–µ –∏–ª–∏ –æ—Ç–∫–ª—é—á–µ–Ω–∏–µ —Ç–∞–π–º–µ—Ä–∞ –ø—Ä–∏ –∏–∑–º–µ–Ω–µ–Ω–∏–∏ —Å–æ—Å—Ç–æ—è–Ω–∏—è –≥–∞–ª–æ—á–∫–∏."""
            if state == QtCore.Qt.Checked:
                self.timer.start(self.current_interval)  # –ó–∞–ø—É—Å–∫–∞–µ–º —Ç–∞–π–º–µ—Ä —Å –≤—ã–±—Ä–∞–Ω–Ω—ã–º –∏–Ω—Ç–µ—Ä–≤–∞–ª–æ–º
            else:
                self.timer.stop()  # –û—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º —Ç–∞–π–º–µ—Ä

    def change_interval(self):
            """–û–±–Ω–æ–≤–ª–µ–Ω–∏–µ –∏–Ω—Ç–µ—Ä–≤–∞–ª–∞ –Ω–∞ –æ—Å–Ω–æ–≤–µ –≤—ã–±—Ä–∞–Ω–Ω–æ–≥–æ –∑–Ω–∞—á–µ–Ω–∏—è –∏–∑ –≤—ã–ø–∞–¥–∞—é—â–µ–≥–æ —Å–ø–∏—Å–∫–∞."""
            selected_text = self.comboBox_time.currentText()
            self.current_interval = self.intervals[selected_text]
            if self.checkBox_auto_search.isChecked():  # –ï—Å–ª–∏ —Ç–∞–π–º–µ—Ä —É–∂–µ –∑–∞–ø—É—â–µ–Ω, –æ–±–Ω–æ–≤–ª—è–µ–º –∏–Ω—Ç–µ—Ä–≤–∞–ª
                self.timer.start(self.current_interval)

    def on_button_click_client(self):
        import socket

        UDP_PORT = 37020
        TIMEOUT = 5

        def discover_server():
            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
            sock.settimeout(TIMEOUT)

            message = b"DISCOVER_SERVER"
            sock.sendto(message, ('<broadcast>', UDP_PORT))
            print("[UDP] –®–∏—Ä–æ–∫–æ–≤–µ—â–∞—Ç–µ–ª—å–Ω—ã–π –∑–∞–ø—Ä–æ—Å –æ—Ç–ø—Ä–∞–≤–ª–µ–Ω. –û–∂–∏–¥–∞–µ–º –æ—Ç–≤–µ—Ç...")

            try:
                data, server_addr = sock.recvfrom(1024)
                if data.decode() == "SERVER_HERE":
                    print(f"[UDP] –°–µ—Ä–≤–µ—Ä –Ω–∞–π–¥–µ–Ω –ø–æ IP: {server_addr[0]}")
                    server_ip = server_addr[0]
                    return server_ip
            except socket.timeout:
                print("[UDP] –í—Ä–µ–º—è –æ–∂–∏–¥–∞–Ω–∏—è –∏—Å—Ç–µ–∫–ª–æ. –°–µ—Ä–≤–µ—Ä –Ω–µ –Ω–∞–π–¥–µ–Ω.")
                return None

        self.counter += 1
        self.label_count_q.setText(f"üì°–ö–æ–ª–∏—á–µ—Å—Ç–≤–æ –æ—Ç–ø—Ä–∞–≤–µ–Ω–Ω—ã—Ö –∑–∞–ø—Ä–æ—Å–æ–≤: {str(self.counter)}")

        server_ip = discover_server()
        if server_ip:
            print(f"[‚úì] IP —Å–µ—Ä–≤–µ—Ä–∞: {server_ip}")
            self.checkBox_auto_search.setChecked(False)
            self.textBrowser_errors.setStyleSheet("color: green;")
            self.textBrowser_errors.setText(f"üñ•Server IP: {server_ip}, ‚úÖResponse: SERVER_HERE")
        else:
            print("[‚úó] –°–µ—Ä–≤–µ—Ä –Ω–µ –æ—Ç–≤–µ—Ç–∏–ª.")
            self.textBrowser_errors.setStyleSheet("color: red;")
            self.textBrowser_errors.setText("‚ùå –°–µ—Ä–≤–µ—Ä –Ω–µ –æ—Ç–≤–µ—Ç–∏–ª.")


if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    main_window = Ui()
    main_window.show()
    sys.exit(app.exec_())
